src/main/java/bio/terra/service/filedata/azure/tables/TableDependencyDao.java

```diff
@@ -0,0 +1,107 @@
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.apache.commons.collections4.ListUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TableDependencyDao {
+  private final Logger logger = LoggerFactory.getLogger(TableDependencyDao.class);
+  private static final String DEPENDENCY_TABLE_NAME_SUFFIX = "dependencies";
+  private static final int MAX_FILTER_CLAUSES = 15;
+
+  @Autowired
+  public TableDependencyDao() {}
+
+  @VisibleForTesting
+  public String getDatasetDependencyTableName(UUID datasetId) {
+    return "datarepo" + datasetId.toString().replaceAll("-", "") + DEPENDENCY_TABLE_NAME_SUFFIX;
+  }
+
+  public void storeSnapshotFileDependencies(
+      TableServiceClient tableServiceClient, UUID datasetId, UUID snapshotId, List<String> refIds) {
+
+    // We construct the snapshot file system without using transactions. We can get away with that,
+    // because no one can access this snapshot during its creation.
+    String dependencyTableName = getDatasetDependencyTableName(datasetId);
+    tableServiceClient.createTableIfNotExists(dependencyTableName);
+    TableClient tableClient = tableServiceClient.getTableClient(dependencyTableName);
+    // The partition size is one less than the MAX_FILTER_CLAUSES to account for the snapshotId
+    // filter
+    ListUtils.partition(refIds, MAX_FILTER_CLAUSES - 1)
+        .forEach(
+            refIdChunk -> {
+              String filter =
+                  refIdChunk.stream()
+                          // maybe wrap or cause in parenthesis
+                          .map(
+                              refId ->
+                                  String.format(
+                                      "%s eq '%s'", FireStoreDependency.FILE_ID_FIELD_NAME, refId))
+                          .collect(Collectors.joining(" or "))
+                      + String.format(
+                          " and %s eq '%s'",
+                          FireStoreDependency.SNAPSHOT_ID_FIELD_NAME, snapshotId);
+              List<TableEntity> entities =
+                  TableServiceClientUtils.filterTable(
+                      tableServiceClient, dependencyTableName, filter);
+              List<String> existing =
+                  entities.stream()
+                      .map(e -> e.getProperty(FireStoreDependency.FILE_ID_FIELD_NAME).toString())
+                      .collect(Collectors.toList());
+              // Create any entities that do not already exist
+              refIdChunk.stream()
+                  .filter(id -> !existing.contains(id))
+                  .forEach(refId -> createDependencyEntity(tableClient, snapshotId, refId));
+              // Update refCount for existing entities
+              entities.forEach(entity -> updateRefCount(tableClient, entity));
```

Comment:
I've been trying to think through this, but I wonder if we could stream line this by using the "upsert" operation. 
-> If it exists, we upsert with the updated refcount
-> If it doesn't exist already, we upsert with ref count =1

I'm worried the existing implementation is going to be brittle b/c we're going to run into the issue with ["createEntity" again](https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java#L104) - It fails if the entry already exists, while "upsert" overwrites the existing entity. This models the behavior of firestore.

Reply:
Thanks for pointing that out! I talked to @nmalfroy yesterday about the `refCount`, and it looks like it will always be 1 since this method now gets passed in a unique list of file ids. We're also not using or updating the `refCount` anywhere else so I'll change this so it just upserts an entity with refCount = 1.